File: full-stack-asset-transfer-guide/applications/conga-cards/src/connect.ts
/*
 * Copyright contributors to the Hyperledgendary Full Stack Asset Transfer Guide project
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as grpc from '@grpc/grpc-js';
import { connect, Gateway, Identity, Signer, signers } from '@hyperledger/fabric-gateway';
import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import { CLIENT_CERT_PATH, GATEWAY_ENDPOINT, HOST_ALIAS, MSP_ID, PRIVATE_KEY_PATH, TLS_CERT_PATH } from './config';

export async function newGrpcConnection(): Promise<grpc.Client> {
    if (TLS_CERT_PATH) {
        const tlsRootCert = await fs.promises.readFile(TLS_CERT_PATH);
        const tlsCredentials = grpc.credentials.createSsl(tlsRootCert);
        return new grpc.Client(GATEWAY_ENDPOINT, tlsCredentials, newGrpcClientOptions());
    }

    return new grpc.Client(GATEWAY_ENDPOINT, grpc.ChannelCredentials.createInsecure());
}

function newGrpcClientOptions(): grpc.ClientOptions {
    const result: grpc.ClientOptions = {};
    if (HOST_ALIAS) {
        result['grpc.ssl_target_name_override'] = HOST_ALIAS; // Only required if server TLS cert does not match the endpoint address we use
    }
    return result;
}

export async function newGatewayConnection(client: grpc.Client): Promise<Gateway> {
    return connect({
        client,
        identity: await newIdentity(),
        signer: await newSigner(),
        // Default timeouts for different gRPC calls
        evaluateOptions: () => {
            return { deadline: Date.now() + 5000 }; // 5 seconds
        },
        endorseOptions: () => {
            return { deadline: Date.now() + 15000 }; // 15 seconds
        },
        submitOptions: () => {
            return { deadline: Date.now() + 5000 }; // 5 seconds
        },
        commitStatusOptions: () => {
            return { deadline: Date.now() + 60000 }; // 1 minute
        },
    });
}

async function newIdentity(): Promise<Identity> {
    const certPath = path.resolve(CLIENT_CERT_PATH);
    const credentials = await fs.promises.readFile(certPath);

    return { mspId: MSP_ID, credentials };
}

async function newSigner(): Promise<Signer> {
    const keyPath = path.resolve(PRIVATE_KEY_PATH);
    const privateKeyPem = await fs.promises.readFile(keyPath);
    const privateKey = crypto.createPrivateKey(privateKeyPem);

    return signers.newPrivateKeySigner(privateKey);
}

File: full-stack-asset-transfer-guide/applications/trader-go/connect.go
package main

import (
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"github.com/hyperledger/fabric-gateway/pkg/client"
	"github.com/hyperledger/fabric-gateway/pkg/hash"
	"github.com/hyperledger/fabric-gateway/pkg/identity"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/insecure"
)

// newGrpcConnection creates a gRPC connection to the Gateway server.
// If TLS_CERT is set, TLS is used; otherwise an insecure connection is established.
func newGrpcConnection() (*grpc.ClientConn, error) {
	endpoint, err := gatewayEndpoint()
	if err != nil {
		return nil, err
	}

	tlsCert := tlsCertPath()
	if tlsCert != "" {
		certPEM, err := os.ReadFile(tlsCert)
		if err != nil {
			return nil, fmt.Errorf("failed to read TLS certificate: %w", err)
		}

		cert, err := identity.CertificateFromPEM(certPEM)
		if err != nil {
			return nil, fmt.Errorf("failed to parse TLS certificate: %w", err)
		}

		certPool := x509.NewCertPool()
		certPool.AddCert(cert)

		transportCreds := credentials.NewClientTLSFromCert(certPool, "")
		opts := []grpc.DialOption{grpc.WithTransportCredentials(transportCreds)}

		// Override TLS server name if endpoint address doesn't match the certificate
		if alias := hostAlias(); alias != "" {
			opts = append(opts, grpc.WithAuthority(alias))
		}

		return grpc.NewClient(endpoint, opts...)
	}

	return grpc.NewClient(endpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
}

// newGatewayConnection creates a Fabric Gateway connection using the provided gRPC connection.
func newGatewayConnection(grpcConn *grpc.ClientConn) (*client.Gateway, error) {
	id, err := newIdentity()
	if err != nil {
		return nil, err
	}

	sign, err := newSigner()
	if err != nil {
		return nil, err
	}

	return client.Connect(
		id,
		client.WithSign(sign),
		client.WithHash(hash.SHA256),
		client.WithClientConnection(grpcConn),
		// Default timeouts for different gRPC calls
		client.WithEvaluateTimeout(5*time.Second),
		client.WithEndorseTimeout(15*time.Second),
		client.WithSubmitTimeout(5*time.Second),
		client.WithCommitStatusTimeout(1*time.Minute),
	)
}

// newIdentity creates a client X.509 identity from the certificate file.
func newIdentity() (*identity.X509Identity, error) {
	certPath, err := clientCertPath()
	if err != nil {
		return nil, err
	}

	certPEM, err := os.ReadFile(certPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read certificate: %w", err)
	}

	cert, err := identity.CertificateFromPEM(certPEM)
	if err != nil {
		return nil, fmt.Errorf("failed to parse certificate: %w", err)
	}

	msp, err := mspID()
	if err != nil {
		return nil, err
	}

	return identity.NewX509Identity(msp, cert)
}

// newSigner creates a signing function from the private key file.
func newSigner() (identity.Sign, error) {
	keyPath, err := privateKeyPath()
	if err != nil {
		return nil, err
	}

	keyPEM, err := os.ReadFile(keyPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read private key: %w", err)
	}

	privateKey, err := identity.PrivateKeyFromPEM(keyPEM)
	if err != nil {
		return nil, fmt.Errorf("failed to parse private key: %w", err)
	}

	return identity.NewPrivateKeySign(privateKey)
}
